// @tarojs/taro-rn/dist/lib/accelerometer.js
import { Accelerometer } from 'expo-sensors';
const accCase = {
    callbacks: [],
};
const intervalMap = {
    game: 20,
    ui: 60,
    normal: 200
};
function offAccelerometerChange(fnc) {
    if (fnc && typeof fnc === 'function') {
        accCase.callbacks = accCase.callbacks.filter((cb) => cb !== fnc);
    }
    else if (fnc === undefined) {
        accCase.callbacks = [];
    }
    else {
        console.warn('offAccelerometerChange failed');
    }
}
function onAccelerometerChange(fnc) {
    accCase.callbacks.push(fnc);
}
/**
 * 开始监听加速度数据。
 * @param {Object} opts
 * @param {string} [opts.interval='normal'] 监听加速度数据回调函数的执行频率
 */
function startAccelerometer(opts = {}) {
    const { interval = 'normal', success, fail, complete } = opts;
    accCase.interval = interval;
    const res = { errMsg: 'startAccelerometer:ok' };
    try {
        // 适配微信小程序行为：重复 start 失败
        if (accCase.listener) {
            throw new Error('startAccelerometer:fail');
        }
        accCase.listener = Accelerometer.addListener((e) => {
            accCase.callbacks.forEach((cb) => {
                cb === null || cb === void 0 ? void 0 : cb(e);
            });
        });
        success === null || success === void 0 ? void 0 : success(res);
        complete === null || complete === void 0 ? void 0 : complete(res);
    }
    catch (error) {
        res.errMsg = 'startAccelerometer:fail';
        fail === null || fail === void 0 ? void 0 : fail(res);
        complete === null || complete === void 0 ? void 0 : complete(res);
        return Promise.reject(res);
    }
    Accelerometer.setUpdateInterval(intervalMap[interval]);
    return Promise.resolve(res);
}
/**
 * 停止监听加速度数据
 * @param opts
 */
function stopAccelerometer(opts = {}) {
    const { success, fail, complete } = opts;
    const res = { errMsg: 'stopAccelerometer:ok' };
    try {
        accCase.listener.remove();
        accCase.listener = null;
        success === null || success === void 0 ? void 0 : success(res);
        complete === null || complete === void 0 ? void 0 : complete(res);
        return Promise.resolve(res);
    }
    catch (error) {
        res.errMsg = 'stopAccelerometer:fail';
        fail === null || fail === void 0 ? void 0 : fail(res);
        complete === null || complete === void 0 ? void 0 : complete(res);
        return Promise.reject(res);
    }
}
export { onAccelerometerChange, offAccelerometerChange, startAccelerometer, stopAccelerometer };
