// @tarojs/router-rn/dist/view/HeadTitle.js
// eslint-disable-next-line no-use-before-define
import * as React from 'react';
import { View, Text } from 'react-native';
import Loading from './Loading';
import { navigationRef } from '../rootNavigation';
export default class HeadTitle extends React.PureComponent {
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    render() {
        var _a, _b;
        // showLoading自定义的参数放在params中
        const params = ((_b = (_a = navigationRef.current) === null || _a === void 0 ? void 0 : _a.getCurrentRoute()) === null || _b === void 0 ? void 0 : _b.params) || {};
        const { showLoading = false } = (params === null || params === void 0 ? void 0 : params.navigateConfig) || {};
        const { label, color, headerProps } = this.props;
        return React.createElement(View, { style: { flexDirection: 'row', alignItems: 'center' } },
            showLoading && React.createElement(Loading, null),
            React.createElement(Text, { style: {
                    fontSize: 17,
                    fontWeight: '600',
                    textAlign: 'center',
                    color: headerProps.tintColor || color,
                    paddingLeft: 2
                } }, (headerProps === null || headerProps === void 0 ? void 0 : headerProps.children) || label));
    }
}
